### PR TITLE
Float funque

### PR DESCRIPTION
**Pull request by ajayshyam**
Initial commit from Ittiam Systems (v1.0)
Includes all functions of Funque implemented in C
float funque is invoked via feature extractor(vmaf_fex_float_funque)

### CODE REVIEW

**Comment by ajayshyam** [comment]
Updated pull request with the following changes:
- Support for API and Interfaces with configurable parameters 
	- Resizer enable / disable 
	- Num VIF scales 
- Added empty objects in the JSON file to maintain consistency with VMAF main branch JSON file.
- Build time configurability for float versus double
- Bug fixes and recent changes
- Submitted as part of commit id (95aa6a6)

File: libvmaf/src/feature/funque_adm.c [hunk 1]
```diff
@@ -0,0 +1,319 @@
+/**
```
Comment (cosmin) [inline libvmaf/src/feature/funque_adm.c]: code related to FUNQUE should be place in `libvmaf/src/feature/third_party/funque/` to keep with the VMAF contribution guide for 3rd party feature extractors

File: libvmaf/meson_options.txt [hunk 2]
```diff
@@ -27,3 +27,13 @@ option('enable_float',
     type: 'boolean',
     value: false,
     description: 'Compile floating-point feature extractors into the library')
+
+option('enable_float_funque',
+    type: 'boolean',
+    value: false,
+    description: 'Compile floating-point funque feature extractors into the library')
+
+option('funque_double_dtype',
```
Comment (cosmin) [inline libvmaf/meson_options.txt]: I think we can simplify this and make double the default, as discussed the float32 usage in the Python version was incidental rather than intentional. We can update the Python code to also use double for parity.

File: libvmaf/src/feature/funque_adm_options.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/* Contrast sensitivity function */
+enum ADM_CSF_MODE {
+    ADM_CSF_MODE_WATSON97 = 0,
+};
+
+/* Default contrast sensitivity function */
+#define DEFAULT_ADM_CSF_MODE (ADM_CSF_MODE_WATSON97)
+
+#define DEFAULT_VIF_LEVELS 2
```
Comment (cosmin) [inline libvmaf/src/feature/funque_adm_options.h]: we should change this to 4 to match core VMAF for now

File: libvmaf/src/feature/funque_adm.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  pyr_rest.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_rest.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[0] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[1] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[2] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+  pyr_add.bands[3] = (funque_dtype *)malloc(sizeof(funque_dtype) * height * width);
+
+  dlm_decouple(ref, dist, dlm_rest, dlm_add);
+
+  dlm_contrast_mask(dlm_rest, dlm_add, pyr_rest, pyr_add, width, height);
```
Comment (abhinaukumar) [inline libvmaf/src/feature/funque_adm.c]: A one-way contrast mask is enough because only `pyr_rest` is used after this step, not `pyr_add`. So, you can also skip allocating memory for `pyr_add`.

File: libvmaf/src/feature/funque_motion.c [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <math.h>
+#include "funque_filters.h"
+/**
+ * Note: img1_stride and img2_stride are in terms of (sizeof(double) bytes)
+ */
+funque_dtype funque_image_sad_c(const funque_dtype *img1, const funque_dtype *img2, int width, int height, int img1_stride, int img2_stride)
```
Comment (abhinaukumar) [inline libvmaf/src/feature/funque_motion.c]: For the sake of readability, calling this `mad` would be more appropriate, since it is the Mean Absolute Difference, not Sum.

File: libvmaf/src/feature/funque_adm.c [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+  funque_dtype *var_k;
+  funque_dtype val;
+  funque_dtype tmp_val;
+
+  for (i = 0; i < height; i++)
+  {
+    for (j = 0; j < width; j++)
+    {
+      index = i * width + j;
+#if FUNQUE_DOUBLE_DTYPE
```
Comment (abhinaukumar) [inline libvmaf/src/feature/funque_adm.c]: As we discussed during the meeting, we should replace the use of `atan` with the cosine-based method. Please refer to https://github.com/Netflix/vmaf/blob/master/libvmaf/src/feature/adm_tools.c#L161.

File: libvmaf/src/feature/funque_vif.c [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+            {
+                sv_sq[index] = var_x[index];
+                g[index] = (double)0;
+            }
+
+            if (sv_sq[index] < exp)
+                sv_sq[index] = exp;
+
+            *score_num += (log((double)1 + g[index] * g[index] * var_x[index] / (sv_sq[index] + sigma_nsq)) + (double)1e-4);
+            *score_den += (log((double)1 + var_x[index] / sigma_nsq) + (double)1e-4);
```
Comment (abhinaukumar) [inline libvmaf/src/feature/funque_vif.c]: You can add a preprocessor flag to switch between the 1e-4 method (used here) and VMAF's method https://github.com/Netflix/vmaf/blob/master/libvmaf/src/feature/vif_tools.c#L369-376. A comparison between the two would be great.

**Review by abhinaukumar — CHANGES_REQUESTED** [review]

File: libvmaf/src/feature/funque_adm.c (hunk 4 above)
Comment (JeevaRajIttiam) [reply libvmaf/src/feature/funque_adm.c]: Thanks for pointing out. This redundancy is removed now.

File: libvmaf/src/feature/funque_motion.c (hunk 5 above)
Comment (ajayshyam) [reply libvmaf/src/feature/funque_motion.c]: The funque_image_sad_c function is renamed to funque_image_mad_c

File: libvmaf/src/feature/funque_adm.c (hunk 1 above)
Comment (ajayshyam) [reply libvmaf/src/feature/funque_adm.c]: All the files related to FUNQUE are moved to libvmaf/src/feature/third_party/funque

File: libvmaf/src/feature/funque_adm.c (hunk 4 above)
Comment (abhinaukumar) [reply libvmaf/src/feature/funque_adm.c]: Thanks @JeevaRajIttiam!

File: libvmaf/src/feature/funque_adm_options.h (hunk 3 above)
Comment (ajayshyam) [reply libvmaf/src/feature/funque_adm_options.h]: Support for 4 levels of VIF is added

Comment (IttiamVijayakumarGR) [reply libvmaf/src/feature/funque_adm_options.h]: VIF, Level 1 and 2 are default

3 and 4 are added based on the num_vif levels configured in json file

File: libvmaf/meson_options.txt (hunk 2 above)
Comment (IttiamVijayakumarGR) [reply libvmaf/meson_options.txt]: Since Float and Double have scores matching upto 5th decimal places float data type has been finalized

Float datatype is used for all buffer (Spatial filter, DWT bands), intermediate squares and accumulators have been double  (in the float version), these buffers have to be double to maintain precession, in Integer implementation double datatype buffers will be converted to int64

File: libvmaf/src/feature/funque_adm.c (hunk 6 above)
Comment (IttiamVijayakumarGR) [reply libvmaf/src/feature/funque_adm.c]: based on the recent test with cosine-based method we have observed maximum difference of 0.007 in corner case, we will mark this issue as closed and can be revisited later on need basis.

**Review by abhinaukumar — APPROVED** [review]
